plugins/net_plugin/include/eosio/net_plugin/protocol.hpp

```diff
@@ -135,15 +135,141 @@ namespace eosio {
+            case both : return "";
+            case peer : return ":peer";
+            default: return ":all";
+        }
+    }
+
+    struct peer_address {
+        peer_address( address_type_enum t = all) : host(), port(), address_type(t), receive(), last_active(), manual(false) {}
+        std::string host;
+        std::string port;
+        address_type_enum address_type;
+        fc::time_point receive;
+        fc::time_point last_active;
+        bool manual;
+
+        // if host and port are same then ignore other configuration
+        bool operator==(const peer_address& other) const {
+            return host == other.host && port == other.port;
+        }
+
+        bool operator!=(const peer_address& other) const {
+            return host != other.host || port != other.port;
+        }
+
+        std::string to_address() const {
+            return host + ":" + port;
+        }
+        std::string to_str() const {
+            return host + ":" + port + address_type_str(address_type);
+        }
+
+        static peer_address from_str(const std::string& input_address_str) {
+            return from_str(input_address_str, false);
+        }
+
+        static peer_address from_str(const std::string& input_address_str, bool is_manual) {
+            std::string address_str = input_address_str;
+            // for "localhost:1234 - 012345" str
+            string::size_type pos = address_str.find(' ');
+            if (pos != std::string::npos) {
+                address_str = address_str.substr(0, pos);
+            }
+            // for "eosproducer1,p2p.eos.io:9876" str
+            pos = address_str.find(',');
+            if (pos != std::string::npos) {
+                address_str = address_str.substr(pos + 1);
+            }
+
+            string::size_type colon = address_str.find(':');
+            string::size_type colon2 = address_str.find(':', colon + 1);
+            string::size_type end = colon2 == string::npos
+                                    ? string::npos : address_str.find_first_of( " :+=.,<>!$%^&(*)|-#@\t", colon2 + 1 );
+            string host_str = address_str.substr( 0, colon );
+            string port_str = address_str.substr( colon + 1, colon2 == string::npos ? string::npos : colon2 - (colon + 1));
+            string type_str = colon2 == string::npos ? "" : end == string::npos ?
+                                                            address_str.substr( colon2 + 1 ) : address_str.substr( colon2 + 1, end - (colon2 + 1) );
+
+            address_type_enum address_type;
+
+            // determine address type
+            if (type_str == "trx") {
+                address_type = trx;
+            } else if (type_str == "blk") {
+                address_type = blk;
+            } else if (type_str == "peer") {
+                address_type = peer;
+            } else if (type_str.empty()) {
+                address_type = both;
+            } else
+                address_type = all;
```

Comment:
I suggest making separate inline utility function to convert string to address type. It will make the code leaner and more readable, plus the function can be reused elsewhere if needed.

Reply:
// determine address type
    inline address_type_enum str_to_address_type(const std::string& address_type_str) {
        static const std::unordered_map<std::string, address_type_enum> address_type_map = {
                {"blk", blk},
                {"trx", trx},
                {"peer", peer},
                {"", both},
                {"all", all},
        };
        const auto it = address_type_map.find(address_type_str);
        if (it != address_type_map.end()) {
            return it->second;
        }
        // if address type not match, return all by default
        return all;
    }

Reply:
LGTM